---
 plugins/faad2/faad2.c |    2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

Index: libquicktime/plugins/faad2/faad2.c
===================================================================
--- libquicktime.orig/plugins/faad2/faad2.c
+++ libquicktime/plugins/faad2/faad2.c
@@ -155,7 +155,7 @@ static int decode_packet_faad2(quicktime
       if(!quicktime_trak_read_packet(file, atrack->track, &codec->pkt))
         return 0;
 
-      memset(&codec->frame_info, 0, sizeof(&codec->frame_info));
+      memset(&codec->frame_info, 0, sizeof(codec->frame_info));
       
       codec->samples = faacDecDecode(codec->dec, &codec->frame_info,
                                      codec->pkt.data, codec->pkt.data_len);
